backend/src/services/external-migration/external-migration-service.ts

```diff
@@ -0,0 +1,185 @@
+import { decryptEnvKeyData, parseEnvKeyData } from "./external-migration-fns";
+import { TImportEnvKeyDataCreate, TImportInfisicalDataCreate } from "./external-migration-types";
+
+type TExternalMigrationServiceFactoryDep = {
+  projectService: TProjectServiceFactory;
+  orgService: TOrgServiceFactory;
+  projectEnvService: TProjectEnvServiceFactory;
+  secretService: TSecretServiceFactory;
+};
+
+export type TExternalMigrationServiceFactory = ReturnType<typeof externalMigrationServiceFactory>;
+
+export const externalMigrationServiceFactory = ({
+  projectService,
+  orgService,
+  projectEnvService,
+  secretService
+}: TExternalMigrationServiceFactoryDep) => {
+  const importInfisicalData = async ({
+    data,
+    actor,
+    actorId,
+    actorOrgId,
+    actorAuthMethod
+  }: TImportInfisicalDataCreate) => {
+    // Import data to infisical
+    if (!data || !data.projects) {
+      logger.error("No projects found in data");
+      return {
+        success: false,
+        message: "No projects found in data"
+      };
+    }
+
+    const orginalToNewProjectId = new Map<string, string>();
+    const orginalToNewEnvironmentId = new Map<string, string>();
+
+    // Import projects
+    const projectPromises = [];
+    for (const [id, project] of data.projects) {
+      const projectPromise = projectService
+        .createProject({
+          actor,
+          actorId,
+          actorOrgId,
+          actorAuthMethod,
+          workspaceName: project?.name,
+          createDefaultEnvs: false
+        })
+        .then((projectResponse) => {
+          if (!projectResponse) {
+            logger.error(`Failed to import project: [name:${project.name}] [id:${id}]`);
+            throw new Error(`Failed to import project: [name:${project.name}] [id:${id}]`);
+          }
+          orginalToNewProjectId.set(project.id, projectResponse.id);
+        });
+      projectPromises.push(projectPromise);
+    }
+    await Promise.all(projectPromises);
+
+    // Invite user importing projects
+    const response = await orgService.inviteUserToOrganization({
+      actorAuthMethod,
+      actorId,
+      actorOrgId,
+      actor,
+      inviteeEmails: [],
+      orgId: actorOrgId,
+      organizationRoleSlug: OrgMembershipRole.NoAccess,
+      projects: Array.from(orginalToNewProjectId.values()).map((project) => {
+        return {
+          id: project,
+          projectRoleSlug: [ProjectMembershipRole.Member]
+        };
+      })
+    });
+    if (!response) {
+      logger.error(`Failed to invite user to projects: [userId:${actorId}]`);
+      return {
+        success: false,
+        message: `Failed to invite user to project: [userId:${actorId}]`
+      };
+    }
+
+    // Import environments
+    if (data.environments) {
+      for (const [id, environment] of data.environments) {
+        try {
+          // TODO: we can create envs parallely once the position constraint is handled differently
+          // eslint-disable-next-line
```

Comment:
Use `for await (const [id, environment] of data.environments)` instead. 

https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Statements/for-await...of